src/SignalR/common/Shared/ClientResultsManager.cs

```diff
@@ -0,0 +1,195 @@
+    {
+        if (_pendingInvocations.TryGetValue(invocationId, out var item))
+        {
+            type = item.Type;
+            return true;
+        }
+        type = null;
+        return false;
+    }
+
+    public Type GetReturnType(string invocationId)
+    {
+        if (TryGetType(invocationId, out var type))
+        {
+            return type;
+        }
+        throw new InvalidOperationException($"Invocation ID '{invocationId}' is not associated with a pending client result.");
+    }
+
+    // Unused, here to honor the IInvocationBinder interface but should never be called
+    public IReadOnlyList<Type> GetParameterTypes(string methodName)
+    {
+        throw new NotImplementedException();
+    }
+
+    // Unused, here to honor the IInvocationBinder interface but should never be called
+    public Type GetStreamItemType(string streamId)
+    {
+        throw new NotImplementedException();
+    }
+
+    // Custom TCS type to avoid the extra allocation that would be introduced if we managed the cancellation separately
+    // Also makes it easier to keep track of the CancellationTokenRegistration for disposal
+    internal sealed class TaskCompletionSourceWithCancellation<T> : TaskCompletionSource<T>
+    {
+        private readonly ClientResultsManager _clientResultsManager;
+        private readonly string _connectionId;
+        private readonly string _invocationId;
+        private readonly CancellationToken _token;
+
+        private CancellationTokenRegistration _tokenRegistration;
+
+        public TaskCompletionSourceWithCancellation(ClientResultsManager clientResultsManager, string connectionId, string invocationId,
+            CancellationToken cancellationToken)
+            : base(TaskCreationOptions.RunContinuationsAsynchronously)
+        {
+            _clientResultsManager = clientResultsManager;
+            _connectionId = connectionId;
+            _invocationId = invocationId;
+            _token = cancellationToken;
+        }
+
+        // Needs to be called after adding the completion to the dictionary in order to avoid synchronous completions of the token registration
+        // not canceling when the dictionary hasn't been updated yet.
+        public void RegisterCancellation()
+        {
+            if (_token.CanBeCanceled)
+            {
+                _tokenRegistration = _token.UnsafeRegister(static o =>
+                {
+                    var tcs = (TaskCompletionSourceWithCancellation<T>)o!;
+                    tcs.SetCanceled();
+                }, this);
+            }
+        }
+
+        public new void SetCanceled()
+        {
+            // TODO: RedisHubLifetimeManager will want to notify the other server (if there is one) about the cancellation
+            // so it can clean up state and potentially forward that info to the connection
+            _ = _clientResultsManager.TryCompleteResult(_connectionId, CompletionMessage.WithError(_invocationId, "Canceled"));
```

Comment:
I don't like that this Task is not awaited. We should at least log any errors. It looks like the Task is sometimes a call to `RedisHubLifetimeManager.PublishAsync`. If there's a reasonable way to cancel any publish calls and await this during teardown, that'd be ideal.